src/Authoring/WinRT.SourceGenerator/AotOptimizer.cs

```diff
@@ -1082,6 +1116,26 @@ internal static void AddVtableAdapterTypeForKnownInterface(ITypeSymbol classType
                 }
             }
 
+            if (classType is INamedTypeSymbol namedType && namedType.MetadataName == "ObservableCollection`1")
+            {
+                // ObservableCollection make use of an internal built-in type as part of its
+                // implementation for INotifyPropertyChanged.  Handling that manually here.
+                var genericInterfaces = new List<string>() { "System.Collections.IList", "System.Collections.IEnumerable" };
+                vtableAttributes.Add(
+                    new VtableAttribute(
+                        "System.Collections.Specialized",
+                        false,
+                        "SingleItemReadOnlyList",
+                        ImmutableArray<TypeInfo>.Empty,
+                        "System.Collections.Specialized.SingleItemReadOnlyList",

```

Comment:
This doesn't seem right. We should not rely on internal implementation details of the BCL. There isn't even a guarantee that all versions of the BCL will have this same internal layout, nor that it won't change at any point in the future. Can you clarify what the issue is here exactly and why the current proposed solution needs to do this in the first place?

Reply:
I agree that is a risk and this was scoped enough that I did a tactical fix.  There isn't anything we are directly doing with the type here other than creating a `vtable` for it and if the signature of the type changes, you will get an `InvalidCastException` which is what folks are seeing today.

The scenario is you have an `ObservableCollection` binded in `XAML` and then you make a modification to it.  This triggers the `INotifyCollectionChanged` handler in `ObservableCollection` which are projected as it is a custom type mapping and the implementation of it passes a `SingleItemReadOnlyList` as the oldItem or so.  Since that isn't a type that the CsWinRT AOT source generator can see or run on, it doesn't get our vtable attribute.  Thereby when it goes across the ABI, `IList` isn't projected and thereby you get a `InvalidCastException`.

The other fix I am looking at possibly is to have a fallback where when the vtable attribute is missing, we only marshal the statically declared interface on the calling function.  But that is going to take longer to get working and do in a way that won't add more to the size in scenarios where it isn't needed.

Reply:
I see, yeah that's reasonable 🤔

> "when the vtable attribute is missing, we only marshal the statically declared interface on the calling function"

Yeah that's what I had in mind too, we'd just marshal as the interface type. If that's planned, that's fine 👍